pkg/chunk/aws/storage_client_s3.go

```diff
@@ -0,0 +1,183 @@
+package aws
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	ot "github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	awscommon "github.com/weaveworks/common/aws"
+	"github.com/weaveworks/common/instrument"
+)
+
+var (
+	s3RequestDuration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "cortex",
+		Name:      "s3_request_duration_seconds",
+		Help:      "Time spent doing S3 requests.",
+		Buckets:   []float64{.025, .05, .1, .25, .5, 1, 2},
+	}, []string{"operation", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(s3RequestDuration)
+}
+
+type s3storageClient struct {
+	storageClient
+	bucketName string
+	S3         s3iface.S3API
+}
+
+// NewS3StorageClient makes a new AWS-backed StorageClient.
+func NewS3StorageClient(cfg StorageConfig, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+	dynamoDB, err := dynamoClientFromURL(cfg.DynamoDB.URL)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.S3.URL == nil {
+		return nil, fmt.Errorf("no URL specified for S3")
+	}
+	s3Config, err := awscommon.ConfigFromURL(cfg.S3.URL)
+	if err != nil {
+		return nil, err
+	}
+	s3Config = s3Config.WithMaxRetries(0) // We do our own retries, so we can monitor them
+	s3Client := s3.New(session.New(s3Config))
+	bucketName := strings.TrimPrefix(cfg.S3.URL.Path, "/")
+
+	client := s3storageClient{
+		storageClient: storageClient{
+			cfg:       cfg.DynamoDBConfig,
+			schemaCfg: schemaCfg,
+			DynamoDB:  dynamoDB,
+		},
+		S3:         s3Client,
+		bucketName: bucketName,
+	}
+	client.queryRequestFn = client.queryRequest
+	client.batchGetItemRequestFn = client.batchGetItemRequest
+	client.batchWriteItemRequestFn = client.batchWriteItemRequest
```

Comment:
Seems like we could roll them in to the struct initialisation above?

Reply:
Yes; this is copied from what was there before: https://github.com/cortexproject/cortex/pull/1063/files#diff-ffc95905deecb289fa4b6ec35477b78aL201